fix(powershell): use Invoke-Expression to pass args

alexsch01 opened the pull request:
Reopening of https://github.com/npm/cli/pull/7458

alexsch01 committed:
fix(powershell): use Invoke-Expression to pass args

Co-authored-by: noseratio <noseratio@gmail.com>

alexsch01 commented:
Would this be a breaking change for NodeJS 22 and NodeJS 24?

alexsch01 committed:
`$MyInvocation.Statement` is undefined in Windows PowerShell

mbtools commented:
Hi Alex, 

Any change here should go along with new tests that prove what's not working (before the change) but working after. This includes executing `npm run xxx` from the command line and `npm run xxx` in a script in `package.json`. 

I have updated `test/bin/windows-shims.js` to cover passing regular and named parameters to scripts. 

[windows-shims.txt](https://github.com/user-attachments/files/20004186/windows-shims.txt)

Please include this in the PR. 

On my machine (pwsh 7.5.0, npm 10.9.2), all parameters are passed correctly with the current `npm.ps1`. 

Do you have other tests that fail? Please add them.

alexsch01 committed:
`windows-shims.js`: cover passing regular and named parameters to scripts

Co-authored-by: @mbtools

alexsch01 commented:
@mbtools thanks! I'll work on new tests for this PR later today

alexsch01 committed:
add tests for comma-separated values

alexsch01 committed:
remove test that doesn't work in command prompt

alexsch01 committed:
add npx tests

alexsch01 committed:
correct test

alexsch01 commented:
@mbtools uhhh npm.ps1 never supported `npm test -- --param1 -p` syntax

alexsch01 commented:
Ok this PR is ready for review

alexsch01 commented:
@mbtools 

> On my machine (pwsh 7.5.0, npm 10.9.2), all parameters are passed correctly with the current `npm.ps1`.

`npm test -- --param1 -p` doesn't work in Windows PowerShell and PowerShell 7.5.1 (pwsh) using npm 10.9.2
- it doesn't pass `--param1` and `-p`
- this PR fixes that

alexsch01 commented:
This PR is ready for review

alexsch01 commented:
Below "**Output**" is from running the `windows-shims.js` from upstream WITHOUT my PR

This should pass before adding my PR changes

---

@mbtools Ok I have an issue running the tests on my Windows machine

Without my PR, running `node .\test\bin\windows-shims.js` gives 2 failures

line number 240 and line number 253 from https://github.com/npm/cli/blob/latest/test/bin/windows-shims.js
- commit 2d1d8d0ef18a10ac7938380884745f1d3c3cb078 if this "latest" file changes

<details>
  <summary><h1>Output</h1></summary>

PS C:\Users\FakeUser\Downloads\cli> node .\test\bin\windows-shims.js
TAP version 14
# Subtest: shim contents
    1..3
    # Subtest: bash
        ok 1 - should be equivalent strictly
        ok 2 - all other changes are m->x
        1..2
    ok 1 - bash # time=4.053ms

    # Subtest: cmd
        ok 1 - should be equivalent strictly
        ok 2 - all other changes are m->x
        1..2
    ok 2 - cmd # time=1.276ms

    # Subtest: pwsh
        ok 1 - should be equivalent strictly
        ok 2 - all other changes are m->x
        1..2
    ok 3 - pwsh # time=2.412ms

ok 1 - shim contents # time=12.605ms

# Subtest: node-gyp
    ok 1 - node-gyp contains env var
    ok 2 - node-gyp contains path
    ok 3 - node-gyp.cmd contains env var
    ok 4 - node-gyp.cmd contains path
    1..4
ok 2 - node-gyp # time=0.827ms

# Subtest: run shims
    1..6
    # Subtest: cmd
        1..2
        not ok 1 - npm.cmd npm
          ---
          diff: |
            --- expected
            +++ actual
            @@ -1,4 +1,4 @@
             Object {
            -  "status": 0,
            -  "stdout": "npm@11.3.0 C:\\Users\\FakeUser\\Downloads\\cli",
            +  "status": 1,
            +  "stdout": "",
             }
          at:
            fileName: test\bin\windows-shims.js
            lineNumber: 240
            columnNumber: 7
            functionName: matchCmd
            isToplevel: true
          stack: |
            matchCmd (test/bin/windows-shims.js:240:7)
            Test.<anonymous> (test/bin/windows-shims.js:263:7)
            Test.<anonymous> (test/bin/windows-shims.js:253:7)
            Object.<anonymous> (test/bin/windows-shims.js:77:3)
          source: "    const result = spawnPath(cmd, [...args, isNpm ? 'help' :
            '--version'], opts)\r

            \r

            \    t.match(result, {\r

            ------^

            \      status: 0,\r

            \      signal: null,\n"
          ...

        not ok 2 - npx.cmd npx
          ---
          diff: |
            --- expected
            +++ actual
            @@ -1,4 +1,4 @@
             Object {
            -  "status": 0,
            -  "stdout": "11.3.0",
            +  "status": 1,
            +  "stdout": "",
             }
          at:
            fileName: test\bin\windows-shims.js
            lineNumber: 240
            columnNumber: 7
            functionName: matchCmd
            isToplevel: true
          stack: |
            matchCmd (test/bin/windows-shims.js:240:7)
            Test.<anonymous> (test/bin/windows-shims.js:264:7)
            Test.<anonymous> (test/bin/windows-shims.js:253:7)
            Object.<anonymous> (test/bin/windows-shims.js:77:3)
          source: "    const result = spawnPath(cmd, [...args, isNpm ? 'help' :
            '--version'], opts)\r

            \r

            \    t.match(result, {\r

            ------^

            \      status: 0,\r

            \      signal: null,\n"
          ...

    not ok 1 - cmd # time=1434.598ms
      ---
      at:
        fileName: test\bin\windows-shims.js
        lineNumber: 253
        columnNumber: 7
        typeName: Test
      source: "\r

        \  for (const { cmd, skip, name, match } of shells) {\r

        \    t.test(name, t => {\r

        ------^

        \      if (skip.reason) {\r

        \        if (skip.fail) {\n"
      ...

    # Subtest: pwsh
        ok 1 - pwsh - not installed # SKIP
        1..1
    ok 2 - pwsh # time=0.61ms

    # Subtest: git bash
        ok 1 - git bash - not installed # SKIP
        1..1
    ok 3 - git bash # time=0.507ms

    # Subtest: user git bash
        ok 1 - user git bash - not installed # SKIP
        1..1
    ok 4 - user git bash # time=1.478ms

    # Subtest: wsl bash
        ok 1 - wsl bash - not installed # SKIP
        1..1
    ok 5 - wsl bash # time=1.499ms

    # Subtest: cygwin bash
        ok 1 - cygwin bash - not installed # SKIP
        1..1
    ok 6 - cygwin bash # time=1.304ms

not ok 3 - run shims # time=1907.767ms
  ---
  at:
    fileName: test\bin\windows-shims.js
    lineNumber: 77
    columnNumber: 3
    typeName: Object
  source: "})\r

    \r

    t.test('run shims', t => {\r

    --^

    \  const path = t.testdir({\r

    \    ...SHIMS,\n"
  ...

1..3
# { total: 17, pass: 10, fail: 2, skip: 5 }
# time=1953.236ms
PS C:\Users\FakeUser\Downloads\cli>

</details>

mbtools commented:
This is not the same `js` as in this PR:

Your Log: `const result = spawnPath(cmd, [...args, isNpm ? 'help' : '--version'], opts)\r`
PR: `const result = spawnPath(cmd, [...args, ...params], opts)`

alexsch01 commented:
> This is not the same `js` as in this PR:
> 
> Your Log: `const result = spawnPath(cmd, [...args, isNpm ? 'help' : '--version'], opts)\r` PR: `const result = spawnPath(cmd, [...args, ...params], opts)`

Right...I meant I'm getting 2 failing tests before adding my PR

mbtools commented:
That's with Windows PowerShell? I'm not sure if this ever was an option (or tested). Here it's about cmd, pwsh, and bash. If we can make it work for the older Windows PowerShell as well, why not?

noseratio commented:
My 2c. If it's possible to make it work for both legacy (v5.0 and below) and modern (v6+) PowerShell with little efforts, that's great. Otherwise, I'd say specifying `#Requires -Version 6.0` should be enough. For Windows people, it's been long overdue to move over to the modern portable PowerShell.

alexsch01 commented:
> That's with Windows PowerShell? I'm not sure if this ever was an option (or tested). Here it's about cmd, pwsh, and bash. If we can make it work for the older Windows PowerShell as well, why not?

Even with cmd I get 2 failures

alexsch01 committed:
small change

alexsch01 committed:
Update windows-shims.js

alexsch01 committed:
Update windows-shims.js

alexsch01 committed:
make less strict

alexsch01 committed:
oops

alexsch01 committed:
`npm.ps1` fix multiple commands

alexsch01 committed:
`npx.ps1` fix multiple commands

alexsch01 commented:
I manually tested that `echo 'test'; npm help a=1,b=2,c=3` now works with this PR (locally not the test) in both Windows PowerShell and pwsh

alexsch01 committed:
Update windows-shims.js

alexsch01 commented:
the 2 failing tests I get on my machine do not appear in GitHub Actions CI so it must be my setup issue

alexsch01 committed:
formatting change

alexsch01 committed:
fix typos

mbtools commented:
Why did you remove the tests with "test --"? Those are important

alexsch01 committed:
replace tests

alexsch01 commented:
> Why did you remove the tests with "test --"? Those are important

just readded those, I couldn't get CI and my local machine to match but now I resolved it

alexsch01 committed:
Update windows-shims.js

alexsch01 committed:
Update windows-shims.js

alexsch01 commented:
@mbtools
UPDATE: tests pass locally with Windows PowerShell instead of pwsh, will have to figure out those differences between the two
I had to escape the double-quotes in the pwsh.exe versions (on my local fork it is powershell.exe) and now it is better

```js
if (cmd.toLowerCase().endsWith('pwsh.exe')) {
    // powershell requires escaping the double-quotes for this test
    args = args.map(elem => elem.replaceAll('"', '\\"'))
}
```

alexsch01 commented:
Other than the other issue...... I'm facing a big issue in GitHub Actions CI - bash can't seem to find the package.json file which is essential for these tests @mbtools

```
                +    npm error enoent Could not read package.json: Error: ENOENT: no such file or directory, open 
'C:\\cygwin64\\home\\runneradmin\\package.json'
```

mbtools commented:
Previous tests didn't check `npm test` i.e. running scripts. It should be looking in `t.testdir` but it's checking `C:\\cygwin64\\home\\runneradmin\\`. It looks like `cygwin bash` doesn't respect `option { cwd: path }` when calling `spawnSync`. 🤔 

You could try `NODE_DEBUG="spawnSync"` to see if the options are passed correctly. After that's it's deep into node. We might have to skip the `npm test` cases for `cygwin`. 🤷

alexsch01 committed:
skip tests that can't pass with cygwin

alexsch01 committed:
more consistent

alexsch01 committed:
more consistent

alexsch01 committed:
revert bad commits

alexsch01 committed:
add back consistency commits, add check for Windows PowerShell but don't use it in CI

alexsch01 commented:
@mbtools

[abc.ps1]
```
echo $args
```

---

[script.js -Windows PowerShell]
```js
const { spawnSync } = require('child_process')

const cmd = String.raw`C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe`
const args = [ 'abc.ps1', 'first a=1,b=2,c=3' ]

const result = spawnSync(`"${cmd}"`, args, {
    env: { PATH: __dirname },
    shell: true,
})

{
    result.stdout = result.stdout?.toString()?.trim()
    result.stderr = result.stderr?.toString()?.trim()
}
console.log(result.stdout)
```

**>node script.js** [Windows PowerShell]
```
first
a=1
b=2
c=3
```

---

[script.js - pwsh7]
```js
const { spawnSync } = require('child_process')

const cmd = String.raw`C:\Program Files\PowerShell\7\pwsh.EXE`
const args = [ 'abc.ps1', 'first a=1,b=2,c=3' ]

const result = spawnSync(`"${cmd}"`, args, {
    env: { PATH: __dirname },
    shell: true,
})

{
    result.stdout = result.stdout?.toString()?.trim()
    result.stderr = result.stderr?.toString()?.trim()
}
console.log(result.stdout)
```

**>node script.js** [pwsh7]
```
first
a=1,b=2,c=3
```

alexsch01 commented:
@mbtools what above means is `pwsh.exe` using spawnSync won't even show the proper difference

mbtools commented:
This looks correct to me. In [pwsh7](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_parsing?view=powershell-7.5), comma is a separator for values of an array. So `a=1,b=2,c=3` is one parameter (with 3 string values `a=1 b=2 c=3`).

alexsch01 commented:
Ok if I run `.\abc.ps1 first a=1,b=2,c=3` in pwsh.exe [without nodejs / npm at all] I get the following output which is different than running it with the above script 

```
first
a=1
b=2
c=3
```

It seems like `spawnSync` sort of escapes the comma with pwsh7, which isn't what I expect

mbtools commented:
It depends on how you pass the parameters to spawn. Args is an array but the test passes just one string (not split by parameters). The reason was testing the part after --. Now there are other tests that might require a split first. 

The way we call spawn here actually must match how npm run script does it. Otherwise, the test results will be misleading. It's a separate package. I can look tomorrow.

alexsch01 commented:
> It depends on how you pass the parameters to spawn. Args is an array but the test passes just one string (not split by parameters). The reason was testing the part after --. Now there are other tests that might require a split first.
> 
> The way we call spawn here actually must match how npm run script does it. Otherwise, the test results will be misleading. It's a separate package. I can look tomorrow.

it's bizarre `spawnSync` with Windows PowerShell gives what I think the correct behavior, where as `spawnSync` with pwsh7 gives what I think the wrong behavior
- when I mean the correct behavior, I mean that it gives the same output for `.\abc.ps1 first a=1,b=2,c=3` and here https://github.com/npm/cli/pull/8267#issuecomment-2849283355

---

For both Windows PowerShell and pwsh7, doing `.\abc.ps1 first a=1,b=2,c=3` gives
```
first
a=1
b=2
c=3
```

alexsch01 commented:
Converted to a draft.....

this will fail the scenario in Windows PowerShell

```js
{
  bin: 'npm',
  params: ['test -- aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa'],
  expected: `aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa`
},
```

alexsch01 committed:
complete change

alexsch01 committed:
also for npx.ps1

alexsch01 committed:
oop

alexsch01 committed:
formatting

alexsch01 commented:
Modifed the code to try to replicate $MyInvocation.Statement on Windows PowerShell

mbtools commented:
I was afraid of this slight mess...

Here's how [`npm run script`](https://github.com/npm/run-script/blob/main/lib/run-script-pkg.js) works. It uses [`@npmcli/promise-spawn`](https://github.com/npm/promise-spawn/blob/main/lib/index.js) which has [this big issue](https://github.com/npm/promise-spawn/issues/73). I have seen this before when porting `pacote` tests to Windows and left [this comment](https://github.com/npm/pacote/pull/408#issuecomment-2453044642).

And here we are again... Calling `spawn` on Windows is a very tricky thing. There are solution like [`cross-spawn`](https://www.npmjs.com/package/cross-spawn) but that's not an option here.

I know you're interested in getting Windows PowerShell to work. Can we get back to the primary goal to solve issues with passing named parameters to PowerShell7 (`$args` doesn't work)?

alexsch01 committed:
remove normalize part

alexsch01 committed:
check value of `$MyInvocation.OffsetInLine`

alexsch01 committed:
finally get it to pass

alexsch01 commented:
I changed the test to add a case specifically for pwsh.exe since it's very different from the rest

@mbtools I also updated the ps1 scripts, I made it so it goes back to `$args` if `$MyInvocation.OffsetInLine` is not valid

alexsch01 committed:
add `powershell.exe` test, have default case for PS1 scripts, revert specific pwsh.exe test

alexsch01 commented:
OK @mbtools, I'm all done. Let me know what you think of it

alexsch01 committed:
Merge branch 'npm:latest' into latest

mbtools commented:
### This is good

We are closer than ever 👍 

![image](https://github.com/user-attachments/assets/fa802879-9763-4fc4-b672-6eeff9998d29)

Maybe we can combine some test cases. I used this one:

`{ bin: 'npm', params: ['test -- -a 1 --b 2 --c=3 "4" --d="5"'], expected: `-a\n1\n--b\n2\n--c=3\n4\n--d=5` },`

### Not so good

The solution fails if there are spaces in the path to `node` or `npm-cli.js`. 

![image](https://github.com/user-attachments/assets/a4decc0d-1df0-450e-b9ef-2a46a0654662)

The full expression looks like this:

`& C:\Program Files\nodejs/node.exe C:\Program Files\nodejs/node_modules/npm/bin/npm-cli.js run test`

It does need the quoting 😉

alexsch01 committed:
make simpler

alexsch01 commented:
Thanks for the feedback! I'll look into the not so good part tomorrow

I also just made the ps1 files simpler by going back to the old logic if the grave key is found in the command....too many edge cases with that one

alexsch01 committed:
oops

alexsch01 committed:
properly quote `$NODE_EXE` and `$NPM_CLI_JS`/`$NPX_CLI_JS` with Invoke-Expression

alexsch01 commented:
Addressed the missing quotes issue

alexsch01 committed:
use reflection to get the full powershell command

wraithgar commented on test/bin/windows-shims.js:
```diff
@@ -259,9 +300,17 @@ t.test('run shims', t => {
+          (bin === 'npm' && params[0].startsWith('test')) ||
+          (bin === 'npx' && params[0].startsWith('.'))
+        )) {
+          t.skip("`cygwin bash` doesn't respect option `{ cwd: path }` when calling `spawnSync`")
```
This ... is troubling.  Do you have more info on this?  If we can't specify what directory to run in we may have to not support cygwin bash.

alexsch01 replied:
context for this was in https://github.com/npm/cli/pull/8267#issuecomment-2849243562 @mbtools

wraithgar replied:
Yeah this isn't a blocker for this PR, it's net new info about cygwin bash.


I created a tracking issue for us to get to this eventually https://github.com/npm/statusboard/issues/980

alexsch01 committed:
use `$MyInvocation.Statement` if exists already

alexsch01 committed:
combine tests

alexsch01 committed:
fix test

alexsch01 commented:
Ok this PR looks complete to me, but if there's anything else to change please let me know

alexsch01 committed:
make simpler

wraithgar commented:
Letting CI run.  Will let @mbtools weigh in here before landing.  Node 24 is now live with npm 11 so it will be good to get this landed.

mbtools commented:
👍 

Let's add single quoted parameters to the test:

`npm run test -- hello -p1 world -p2 "hello world" -p3 'hello world' --q1=hello world --q2="hello world" --q3='hello world'`

```
hello
-p1
world
-p2
hello world
-p3
hello world
--q1=hello
world
--q2=hello world
--q3=hello world
```

I'm not sure how you came up with the $NPM_OG_COMMAND syntax but it works like a charm. Thank you!

Just a bit of styling to keep "exit" at the end:

```ps
if { 
  #...

  # Support pipeline input
  if ($MyInvocation.ExpectingInput) {
    $input | Invoke-Expression "& `"$NODE_EXE`" `"$NPM_CLI_JS`" $NPM_ARGS"
  } else {
    Invoke-Expression "& `"$NODE_EXE`" `"$NPM_CLI_JS`" $NPM_ARGS"
  }
} else {
  # Support pipeline input
  if ($MyInvocation.ExpectingInput) {
    $input | & $NODE_EXE $NPM_CLI_JS $args
  } else {
    & $NODE_EXE $NPM_CLI_JS $args
  }
}

exit $LASTEXITCODE
```

alexsch01 committed:
single exit for ps1 bin files, single-quote test

alexsch01 committed:
Merge branch 'npm:latest' into latest

mbtools commented:
with single quotes, it works on the command line but fails the spawn test with `cmd` and `pwsh`:

![image](https://github.com/user-attachments/assets/5a277b68-4a11-4dea-aa1c-34417c4019e8)

alexsch01 commented:
@mbtools 

CMD treats single-quotes as literals - output on my machine

```
hello
-p1
world
-p2
hello world
-p3
'hello
world'
--q1=hello
world
--q2=hello world
--q3='hello
world'
```

Don't know why pwsh7 is not passing the test

alexsch01 commented:
It almost feels like pwsh7 and cmd output always match with spawnSync ??

```js
const { spawnSync } = require('child_process')

const cmd = 'pwsh'
const result = spawnSync(`"${cmd}"`, [`.\\abc.ps1 "hello world" 'hi world' a=1,b=2,c=3`], {
	shell: true,
        cwd: __dirname,
})

if(result.stderr.length > 0) throw new Error(result.stderr.toString().trim())
console.log(result.stdout.toString().trim())
```

```
>node script.js
hello world
'hi
world'
a=1,b=2,c=3
```

---

```js
const { spawnSync } = require('child_process')

const cmd = 'powershell'
const result = spawnSync(`"${cmd}"`, [`.\\abc.ps1 "hello world" 'hi world' a=1,b=2,c=3`], {
	shell: true,
        cwd: __dirname,
})

if(result.stderr.length > 0) throw new Error(result.stderr.toString().trim())
console.log(result.stdout.toString().trim())
```

```
>node script.js
hello
world
hi world
a=1
b=2
c=3
```

mbtools commented:
Let's remove the single quote tests for now. It doesn't look related to the `ps1` scripts. We will figure it out after (new issue).

This PR is hard to parse now. Could you create a new PR with a single commit and summary of the change (co-pilot 😉), please?

alexsch01 commented:
Continuing in https://github.com/npm/cli/pull/8278

mbtools commented:
PS: Set `$env:NODE_DEBUG="*"` and run `node script.js`. Then you see how spawn will map it back to `cmd.exe` and sets `args` (due to  `shell: true`). So in the end it's always `cmd.exe`.

alexsch01 commented:
> PS: Set `$env:NODE_DEBUG="*"` and run `node script.js`. Then you see how spawn will map it back to `cmd.exe` and sets `args` (due to `shell: true`). So in the end it's always `cmd.exe`.

It seems like pwsh works differently from powershell when it's done this way

`cmd /d /s /c "pwsh .\abc.ps1 "hello world" 'hi world' a=1,b=2,c=3"`

vs

`cmd /d /s /c "powershell .\abc.ps1 "hello world" 'hi world' a=1,b=2,c=3"`

alexsch01 commented:
ahhhh `cmd /d /s /c "pwsh -Command .\abc.ps1 "hello world" 'hi world' a=1,b=2,c=3"` gives the expected output

```
hello
world
hi world
a=1
b=2
c=3
```

EDIT: I updated the new PR

alexsch01 commented:
@mbtools with the single-quote tests which I'm not merging into my branch.....we now get this https://github.com/alexsch01/cli/actions/runs/14887289051/job/41810161408

Only CMD is now failing!!! which is expected for CMD

ehoogeveen-medweb commented:
Yeah, `cmd /d /s /c "pwsh -Command .\abc.ps1 "hello world" 'hi world' a=1,b=2,c=3"` should evaluate `pwsh -Command .\abc.ps1 "hello world" 'hi world' a=1,b=2,c=3` which would pass `"hello world"` as a single argument but split `'hi world'` into `'hi` and `world'` since CMD has no special handling for single quotes, right? You said it's expected, so just confirming since you seemed to report success in your previous comment.

alexsch01 commented:
> Yeah, `cmd /d /s /c "pwsh -Command .\abc.ps1 "hello world" 'hi world' a=1,b=2,c=3"` should evaluate `pwsh -Command .\abc.ps1 "hello world" 'hi world' a=1,b=2,c=3` which would pass `"hello world"` as a single argument but split `'hi world'` into `'hi` and `world'` since CMD has no special handling for single quotes, right? You said it's expected, so just confirming since you seemed to report success in your previous comment.

yes the test is good that's what I meant
We aren't going to use the single-quote test
